dbms/src/Encryption/CSEDataKeyManager.cpp

```diff
@@ -0,0 +1,146 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Encryption/CSEDataKeyManager.h>
+#include <IO/MemoryReadWriteBuffer.h>
+#include <Poco/Path.h>
+#include <Storages/KVStore/FFI/FileEncryption.h>
+#include <Storages/KVStore/FFI/ProxyFFI.h>
+#include <Storages/Page/V3/Universal/UniversalPageStorage.h>
+#include <Storages/Page/V3/Universal/UniversalWriteBatchImpl.h>
+#include <openssl/md5.h>
+
+
+namespace DB
+{
+
+CSEDataKeyManager::CSEDataKeyManager(
+    EngineStoreServerWrap * tiflash_instance_wrap_,
+    std::weak_ptr<UniversalPageStorage> & ps_write_)
+    : tiflash_instance_wrap{tiflash_instance_wrap_}
+    , ps_write{ps_write_}
+    , master_key(std::make_unique<MasterKey>(tiflash_instance_wrap->proxy_helper->getMasterKey()))
+{}
+
+FileEncryptionInfo CSEDataKeyManager::getInfo(const EncryptionPath & ep)
+{
+    const auto keyspace_id = ep.keyspace_id;
+    // use likely to reduce performance impact on keyspaces without enable encryption
+    if likely (keyspace_id == NullspaceID || !tiflash_instance_wrap->proxy_helper->getKeyspaceEncryption(keyspace_id))
+    {
+        return FileEncryptionInfo{
+            FileEncryptionRes::Disabled,
+            EncryptionMethod::Plaintext,
+            nullptr,
+            nullptr,
+            nullptr,
+        };
+    }
+
+    auto it = keyspace_id_to_key.find(keyspace_id);
+    if (it == keyspace_id_to_key.end())
+    {
+        // Read encrypted key from PageStorage
+        const auto page_id = UniversalPageIdFormat::toFullPageId(
+            UniversalPageIdFormat::toFullPrefix(keyspace_id, StorageType::Meta, ENCRYPTION_KEY_RESERVED_NAMESPACE_ID),
+            ENCRYPTION_KEY_RESERVED_PAGEU64_ID);
+        // getFile will be called after newFile, so page must exist
+        Page page = ps_write.lock()->read(
```

Comment:
Why `ps_write` is a weak_ptr but you don't check whether it is nullptr after each lock?
What's the reason of using weak_ptr instead of shared_ptr?

Reply:
mistake, should use shared_ptr here.